src/midi/midi_mt32.cpp

```diff
@@ -0,0 +1,516 @@
+ *  Copyright (C) 2020       Nikos Chantziaras <realnc@gmail.com> (settings)
+ *  Copyright (C) 2020       The dosbox-staging team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "midi_mt32.h"
+
+#if C_MT32EMU
+
+#include <SDL_thread.h>
+#include <SDL_endian.h>
+
+#include "control.h"
+#include "string_utils.h"
+
+#ifndef DOSBOX_MIDI_H
+#include "midi.h"
+#endif
+
+// Global Constants
+// ----------------
+constexpr uint16_t MS_PER_S = 1000;
+constexpr uint8_t CH_PER_FRAME = 2; // left and right channels
+
+MidiHandler_mt32 mt32_instance;
+
+static void init_mt32_dosbox_settings(Section_prop &sec_prop)
+{
+	constexpr auto when_idle = Property::Changeable::WhenIdle;
+
+	auto *str_prop = sec_prop.Add_string("romdir", when_idle, "");
+	str_prop->Set_help(
+	        "Name of the directory where MT-32 Control and PCM ROM files can be found. Emulation requires these files to work.\n"
+	        "  Accepted file names are as follows:\n"
+	        "    MT32_CONTROL.ROM or CM32L_CONTROL.ROM - control ROM file.\n"
+	        "    MT32_PCM.ROM or CM32L_PCM.ROM - PCM ROM file.");
+
+	auto *bool_prop = sec_prop.Add_bool("reverse.stereo", when_idle, false);
+	bool_prop->Set_help("Reverse stereo channels for MT-32 output");
+
+	bool_prop = sec_prop.Add_bool("verbose", when_idle, false);
+	bool_prop->Set_help("MT-32 debug logging");
+
+	bool_prop = sec_prop.Add_bool("thread", when_idle, false);
+	bool_prop->Set_help("MT-32 rendering in separate thread");
+
+	auto *int_prop = sec_prop.Add_int("chunk", when_idle, 16);
+	int_prop->SetMinMax(2, 100);
+	int_prop->Set_help(
+	        "Minimum milliseconds of data to render at once. (min 2, max 100)\n"
+	        "Increasing this value reduces rendering overhead which may improve performance but also increases audio lag.\n"
+	        "Valid for rendering in separate thread only.");
+
+	int_prop = sec_prop.Add_int("prebuffer", when_idle, 32);
+	int_prop->SetMinMax(3, 200);
+	int_prop->Set_help(
+	        "How many milliseconds of data to render ahead. (min 3, max 200)\n"
+	        "Increasing this value may help to avoid underruns but also increases audio lag.\n"
+	        "Cannot be set less than or equal to mt32.chunk value.\n"
+	        "Valid for rendering in separate thread only.");
```

Comment:
Implementation detail, not to be exposed to users (we already have mixer prebuffer exposed - value in here should be either hardcoded or calculated based on mixer prebuffer).